### DISCUSSION TITLE
Is it possible to have npm support alternative js runtimes besides node?

### DISCUSSION THREAD

**Discussion by ghost** [npm/feedback#661] [CLI] [answered]
Simple question, npm is labelled as an all round javascript package manager and doesn't explicitly say it's only for node. ofcourse in practice this isn't true, but would it be difficult to enhance it to run on any arbitrary js runtime compliant to the spec? this couldbe quite useful in my view

**Comment by ljharb** [comment]
JS doesn’t have any way to access the filesystem or do any i/o, including to the terminal or the network - that’s part of node itself. npm could certainly be written to run on another specific JS runtime, but not on an arbitrary one.

What would the use be?

**Comment by MylesBorins** [comment, marked as answer]
Getting the npm CLI itself to run on other runtimes will not be an easy feat as it is built on top of various Node.js APIs. It definitely would be interesting to explore how it could run on top of other runtimes... but TBH I think it would make far more sense to build new minimal clients that can speak to the registry than make the npm cli itself more generic.

One thing that could be a smaller task would be looking at making some of the building blocks of npm more generic or at least pluggable for system I/O APIs, arborist is the first package I can think of that would be very useful for that.

If you would like to propose anything more concrete it would make sense to open an issue over at https://github.com/npm/rfcs to kick off discussion
